libfreerdp/core/gateway/rdg.c

```diff
@@ -75,57 +75,61 @@ static BOOL rdg_write_packet(rdpRdg* rdg, wStream* sPacket)
 	return TRUE;
 }
 
-static wStream* rdg_receive_packet(rdpRdg* rdg)
+static BOOL rdg_read_all(rdpTls* tls, BYTE* buffer, int size)
 {
 	int status;
-	wStream* s;
-	RdgPacketHeader* packet;
-	UINT32 readCount = 0;
-	s = Stream_New(NULL, 1024);
-
-	if (!s)
-		return NULL;
-
-	packet = (RdgPacketHeader*) Stream_Buffer(s);
+	int readCount = 0;
 
-	while (readCount < sizeof(RdgPacketHeader))
+	while (readCount < size)
 	{
-		status = BIO_read(rdg->tlsOut->bio, Stream_Pointer(s), sizeof(RdgPacketHeader) - readCount);
+		status = BIO_read(tls->bio, buffer, size - readCount);
 
-		if (status < 0)
+		if (status <= 0)
 		{
+			if (!BIO_should_retry(tls->bio))
+				return FALSE;
+
 			continue;
 		}
 
 		readCount += status;
-		Stream_Seek(s, readCount);
 	}
+	return TRUE;
+}
 
-	if (Stream_Capacity(s) < packet->packetLength)
-	{
-		if (!Stream_EnsureCapacity(s, packet->packetLength))
-		{
-			Stream_Free(s, TRUE);
-			return NULL;
-		}
+static wStream* rdg_receive_packet(rdpRdg* rdg)
+{
+	wStream* s;
+	size_t packetLength;
 
-		packet = (RdgPacketHeader*) Stream_Buffer(s);
-	}
+	s = Stream_New(NULL, 1024);
+
+	if (!s)
+		return NULL;
 
-	while (readCount < packet->packetLength)
+	if (!rdg_read_all(rdg->tlsOut, Stream_Buffer(s), sizeof(RdgPacketHeader)))
 	{
-		status = BIO_read(rdg->tlsOut->bio, Stream_Pointer(s), packet->packetLength - readCount);
+		Stream_Free(s, TRUE);
+		return NULL;
+	}
 
-		if (status < 0)
-		{
-			continue;
-		}
+	packetLength = ((RdgPacketHeader*)Stream_Buffer(s))->packetLength;
```

Comment:
Better avoid direct buffer to struct casts (endianess issues ahead)

Reply:
So the alternative would be something like this?
```
Stream_Seek(s, 4);
Stream_Read_UINT32(s, packetLength);
```

Reply:
@p-pautov yes definitely